File: ios/DailyMemory/Data/Remote/EmbeddingService.swift
import Foundation

/// Service for generating text embeddings using OpenAI API
/// Uses text-embedding-3-small model for cost-effective semantic search
actor EmbeddingService {
    static let shared = EmbeddingService()

    private let apiClient = APIClient.shared
    private let dimensions = 1536

    // MARK: - Cache
    private var embeddingCache: [String: [Float]] = [:]
    private let maxCacheSize = 1000

    private init() {}

    // MARK: - Embedding Generation

    /// Generate embedding via backend API
    func generateEmbedding(for text: String) async -> Result<[Float], Error> {
        let cacheKey = String(text.prefix(200).lowercased())
        if let cached = embeddingCache[cacheKey] {
            return .success(cached)
        }

        do {
            let response: EmbeddingResponse = try await apiClient.post("embeddings", body: ["text": text])
            let embedding = response.embedding

            if embeddingCache.count >= maxCacheSize {
                embeddingCache.removeAll()
            }
            embeddingCache[cacheKey] = embedding

            return .success(embedding)
        } catch {
            return .success(simulateEmbedding(for: text))
        }
    }

    /// Store embedding on server for semantic search
    func storeEmbedding(memoryId: String, text: String) async -> Result<Bool, Error> {
        do {
            let _: StoreEmbeddingResponse = try await apiClient.post("embeddings/store", body: [
                "memoryId": memoryId,
                "text": text
            ])
            return .success(true)
        } catch {
            return .failure(error)
        }
    }

    /// Generate embeddings for multiple texts (batch) via backend API
    func generateEmbeddings(for texts: [String]) async -> Result<[[Float]], Error> {
        do {
            struct BatchResponse: Decodable { let embeddings: [[Float]]; let count: Int }
            let response: BatchResponse = try await apiClient.post("embeddings/batch", body: ["texts": texts])
            return .success(response.embeddings)
        } catch {
            let embeddings = texts.map { simulateEmbedding(for: $0) }
            return .success(embeddings)
        }
    }

    // MARK: - Legacy (removed - now uses backend API)

    // MARK: - Simulation

    /// Simulate embedding for testing/demo purposes
    /// Uses simple word-based hashing to create pseudo-embeddings
    private func simulateEmbedding(for text: String) -> [Float] {
        var embedding = [Float](repeating: 0.0, count: dimensions)
        let words = text.lowercased().split(separator: " ")

        for (index, word) in words.enumerated() {
            let hash = word.hashValue
            let position = abs(hash) % dimensions

            // Distribute word influence across multiple dimensions
            for offset in 0..<10 {
                let idx = (position + offset) % dimensions
                let value = Float(hash % 1000) / 1000.0 * (index % 2 == 0 ? 1 : -1)
                embedding[idx] += value / Float(words.count)
            }
        }

        // Normalize the vector
        let magnitude = sqrt(embedding.reduce(0) { $0 + $1 * $1 })
        if magnitude > 0 {
            embedding = embedding.map { $0 / magnitude }
        }

        return embedding
    }

    // MARK: - Cache Management

    func clearCache() {
        embeddingCache.removeAll()
    }

    func getCacheSize() -> Int {
        return embeddingCache.count
    }
}

// MARK: - Errors

enum EmbeddingError: LocalizedError {
    case emptyResponse
    case invalidResponse
    case apiError(statusCode: Int)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .emptyResponse:
            return "Empty response from embedding API"
        case .invalidResponse:
            return "Invalid response from embedding API"
        case .apiError(let statusCode):
            return "Embedding API error: HTTP \(statusCode)"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        }
    }
}

File: android/app/src/main/java/com/dailymemory/data/remote/EmbeddingService.kt
package com.dailymemory.data.remote

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.sqrt

/**
 * Service for generating text embeddings using OpenAI API
 * Uses text-embedding-3-small model for cost-effective semantic search
 */
@Singleton
class EmbeddingService @Inject constructor() {

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    // Configuration
    private var apiKey: String = ""
    private val model = "text-embedding-3-small"
    private val dimensions = 1536

    // Cache
    private val embeddingCache = mutableMapOf<String, FloatArray>()
    private val maxCacheSize = 1000

    /**
     * Configure the service with OpenAI API key
     */
    fun configure(apiKey: String) {
        this.apiKey = apiKey
    }

    /**
     * Generate embedding for a single text
     */
    suspend fun generateEmbedding(text: String): Result<FloatArray> = withContext(Dispatchers.IO) {
        // Check cache first
        val cacheKey = text.take(200).lowercase()
        embeddingCache[cacheKey]?.let {
            return@withContext Result.success(it)
        }

        if (apiKey.isEmpty()) {
            return@withContext Result.success(simulateEmbedding(text))
        }

        try {
            val embedding = fetchEmbedding(text)

            // Cache the result
            if (embeddingCache.size >= maxCacheSize) {
                embeddingCache.clear()
            }
            embeddingCache[cacheKey] = embedding

            Result.success(embedding)
        } catch (e: Exception) {
            // Fallback to simulation on error
            Result.success(simulateEmbedding(text))
        }
    }

    /**
     * Generate embeddings for multiple texts (batch)
     */
    suspend fun generateEmbeddings(texts: List<String>): Result<List<FloatArray>> = withContext(Dispatchers.IO) {
        if (apiKey.isEmpty()) {
            return@withContext Result.success(texts.map { simulateEmbedding(it) })
        }

        try {
            val embeddings = fetchEmbeddings(texts)
            Result.success(embeddings)
        } catch (e: Exception) {
            Result.success(texts.map { simulateEmbedding(it) })
        }
    }

    private fun fetchEmbedding(text: String): FloatArray {
        return fetchEmbeddings(listOf(text)).first()
    }

    private fun fetchEmbeddings(texts: List<String>): List<FloatArray> {
        val requestBody = JSONObject().apply {
            put("model", model)
            put("input", JSONArray(texts))
            put("dimensions", dimensions)
        }.toString()

        val request = Request.Builder()
            .url("https://api.openai.com/v1/embeddings")
            .addHeader("Content-Type", "application/json")
            .addHeader("Authorization", "Bearer $apiKey")
            .post(requestBody.toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: throw EmbeddingException("Empty response")

        if (!response.isSuccessful) {
            throw EmbeddingException("API error: HTTP ${response.code}")
        }

        return parseEmbeddingResponse(responseBody)
    }

    private fun parseEmbeddingResponse(responseBody: String): List<FloatArray> {
        val json = JSONObject(responseBody)
        val dataArray = json.getJSONArray("data")

        // Parse and sort by index
        val embeddings = mutableListOf<Pair<Int, FloatArray>>()

        for (i in 0 until dataArray.length()) {
            val item = dataArray.getJSONObject(i)
            val index = item.getInt("index")
            val embeddingArray = item.getJSONArray("embedding")

            val floatArray = FloatArray(embeddingArray.length()) { j ->
                embeddingArray.getDouble(j).toFloat()
            }

            embeddings.add(index to floatArray)
        }

        return embeddings.sortedBy { it.first }.map { it.second }
    }

    /**
     * Simulate embedding for testing/demo purposes
     */
    private fun simulateEmbedding(text: String): FloatArray {
        val embedding = FloatArray(dimensions) { 0f }
        val words = text.lowercase().split(" ")

        words.forEachIndexed { index, word ->
            val hash = word.hashCode()
            val position = kotlin.math.abs(hash) % dimensions

            // Distribute word influence across multiple dimensions
            for (offset in 0 until 10) {
                val idx = (position + offset) % dimensions
                val value = (hash % 1000) / 1000f * (if (index % 2 == 0) 1 else -1)
                embedding[idx] += value / words.size
            }
        }

        // Normalize the vector
        val magnitude = sqrt(embedding.sumOf { (it * it).toDouble() }).toFloat()
        if (magnitude > 0) {
            for (i in embedding.indices) {
                embedding[i] /= magnitude
            }
        }

        return embedding
    }

    /**
     * Clear the embedding cache
     */
    fun clearCache() {
        embeddingCache.clear()
    }

    /**
     * Get current cache size
     */
    fun getCacheSize(): Int = embeddingCache.size
}

/**
 * Exception for embedding errors
 */
class EmbeddingException(message: String) : Exception(message)
